### ISSUE TITLE
Investigate deprecation of --compilers

### ISSUE THREAD

**Issue by boneskull** [mochajs/mocha#2493]
`--compilers` may be useless with later versions of Node.js, and commonly `--require` is sufficient instead. Example:

```
--compilers js:babel-register
```

Is better written as

```
--require babel-register
```

I don't know if _this_

```
--compilers coffee:coffee-script/register
```

Is equivalent to this:

```
--require coffee-script
```

If so, it's probably only true for some versions of Node.js and/or `coffee-script`. 

If you use a different extension with Babel--say `.es6`--I think you would need `--compilers`? How does this work with TypeScript?

Regardless, `--compilers js:some_module` may always be better written with `--require` if that extension is `.js`.  Something similar is noted in the docs.

**Comment by ScottFreeCode** [comment]
I just looked at the code, and it doesn't seem like Mocha is actually doing anything to associate the particular compiler with the particular extension, just `require`ing the compiler and adding the extension to the list of test/watch extensions, which I believe is used somewhere in the file lookup logic if you haven't specified a glob. In other words, if I understand correct, `--compilers <ext>:<compiler> <folder>` should be equivalent to `--require <compiler> <folder>/*.{js,<ext>}` and `--compilers <ext>:<compiler> --recursive <folder>` should be equivalent to `--require <compiler> <folder>/**/*.{js,<ext>}`. With `--watch` I'm not sure if there's an equivalent (since I'm not aware of any way to specify which files/folders to watch), but considering we would really like to get Mocha to where we can stop having to handle such stuff in `--watch` too...

**Comment by ScottFreeCode** [comment]
I just double-checked control of `watch`'s extensions, and there is indeed another commandline argument for that. `--compilers <ext>:<compiler> --watch <folder>` should be equivalent to `--require <compiler> --watch-extensions <ext> --watch "<folder>/*.{js,<ext>}"` and `--compilers <ext>:<compiler> --watch --recursive <folder>` should be equivalent to `--require <compiler> --watch-extensions <ext> --watch "<folder>/**/*.{js,<ext>}"`

As far as I can see, we should go ahead and document that `--compilers` is just a shortcut for `--require` and adding extensions to your test files pattern (and to `--watch-extensions` if using `--watch`), and then we can get rid of `--compilers` in a semver major version.

**Comment by fesebuv** [comment]
@ScottFreeCode is this issue resolved? Does it require further action?

**Comment by ScottFreeCode** [comment]
As far as I recall, we figured out how to entirely replace `--compilers` with `--require`, `--watch-extensions` and globbing, but I don't remember whether we verified that the needed glob pattern can be quoted with the same syntax on both Unixy and non-Unixy OSes (:cough-windows-cough:), and I'm fairly sure we didn't decide whether to deprecate it or merely document what it's shorthand for. I will double-check that the relevant glob syntax is useable portably and if so ping the team for a decision on deprecation.

**Comment by fesebuv** [comment]
@ScottFreeCode excellent!

**Comment by boneskull** [comment]
Regarding [`.mjs`](https://github.com/nodejs/node-eps/blob/master/002-es-modules.md#32-determining-if-source-is-an-es-module), see [@std/esm](https://npm.im/@std/esm), which is pretty awesome.

**Comment by ScottFreeCode** [comment]
Created #3005 to prove with testing that one can glob multiple extensions, works in both OSes.

The glob has changed since my prior investigation, due to upgrading the Glob package; it is now `@(first_extension|second_extension)` instead of `{first_extension,second_extension}`.

**Comment by boneskull** [comment]
closed via #3038
